### PR TITLE
feat: init binding and async

### PR DESCRIPTION
**Pull request by kehuili**
This big initial PR is an important upgrade to current GCP Python Function Framework with following task accomplished:

- Add OpenFunction knative runtime
- Add OpenFunction async runtime
- Add e2e tests for knative runtime
- Add e2e tests for async runtime
- Add sequence diagram for knative runtime
- Add sequence diagram for async runtime
- Refine README

### CODE REVIEW

**Comment by benjaminhuo** [comment]
> [CLA assistant check](https://cla-assistant.io/OpenFunction/functions-framework-python?pullRequest=1) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/OpenFunction/functions-framework-python?pullRequest=1) before we can accept your contribution.**0** out of **2** committers have signed the CLA.❌ Kehui Li❌ anniefu
> 
> **Kehui Li** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).
> You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/OpenFunction/functions-framework-python?pullRequest=1) it.

That's a great enhancement! Thanks @kehuili 
Would you sign the CLA? You might need to get around the GFW to sign it.

And to utilize this python functions-framework, we might need to update https://github.com/OpenFunction/builder as well.
Let's discuss this in today's meeting.

Ben

**Comment by kehuili** [comment]
> > [CLA assistant check](https://cla-assistant.io/OpenFunction/functions-framework-python?pullRequest=1) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/OpenFunction/functions-framework-python?pullRequest=1) before we can accept your contribution.**0** out of **2** committers have signed the CLA.❌ Kehui Li❌ anniefu
> > **Kehui Li** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).
> > You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/OpenFunction/functions-framework-python?pullRequest=1) it.
> 
> That's a great enhancement! Thanks @kehuili Would you sign the CLA? You might need to get around the GFW to sign it.
> 
> And to utilize this python functions-framework, we might need to update https://github.com/OpenFunction/builder as well. Let's discuss this in today's meeting.
> 
> Ben

Just signed CLA. Also I picked a commit from GCP for a test fix (or one unit test would fail forever), it seems he/she also need to sign. Do I need to revert this commit and just commit this change myself?

**Comment by benjaminhuo** [comment]
> > > [CLA assistant check](https://cla-assistant.io/OpenFunction/functions-framework-python?pullRequest=1) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/OpenFunction/functions-framework-python?pullRequest=1) before we can accept your contribution.**0** out of **2** committers have signed the CLA.❌ Kehui Li❌ anniefu
> > > **Kehui Li** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).
> > > You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/OpenFunction/functions-framework-python?pullRequest=1) it.
> > 
> > 
> > That's a great enhancement! Thanks @kehuili Would you sign the CLA? You might need to get around the GFW to sign it.
> > And to utilize this python functions-framework, we might need to update https://github.com/OpenFunction/builder as well. Let's discuss this in today's meeting.
> > Ben
> 
> Just signed CLA. Also I picked a commit from GCP for a test fix (or one unit test would fail forever), it seems he/she also need to sign. Do I need to revert this commit and just commit this change myself?

No, just you sign the CLA is all right

File: src/openfunction/function_context.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            self.metadata
+        )
+
+    @staticmethod
+    def from_json(json_dct):
+        uri = json_dct.get('uri') or ''
+        component_name = json_dct.get('componentName') or ''
+        metadata = json_dct.get('metadata')
+        component_type = json_dct.get('componentType') or ''
+        operation = json_dct.get('operation') or ''
```
Comment (tpiperatgod) [inline src/openfunction/function_context.py]: Would it be better to use `get("key", "default")`?

**Comment by tpiperatgod** [comment]
lgtm

**Comment by lizzzcai** [comment]
is the quick start example in the [readme](https://github.com/kehuili/functions-framework-python/tree/feature/init_binding_and_async#quickstarts) need to be updated as well?

**Comment by benjaminhuo** [comment]
Kehui Li seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).
You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/OpenFunction/functions-framework-python?pullRequest=1) it.

@kehuili you may need to add your email used to sign this commit as your github primary email and sign the cla again.

https://docs.github.com/cn/account-and-profile/setting-up-and-managing-your-github-user-account/managing-email-preferences/adding-an-email-address-to-your-github-account

File: src/openfunction/function_context.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
+OPEN_FUNC_BINDING = "bindings"
+OPEN_FUNC_TOPIC = "pubsub"
+
+KNATIVE_RUNTIME_TYPE = "Knative"
```
Comment (benjaminhuo) [inline src/openfunction/function_context.py]: Is this runtime type used as the `runtime` field of serving?
If so it should be in lower case `knative` or `async` like https://github.com/OpenFunction/samples/blob/main/functions/async/logs-handler-function/logs-handler-function.yaml#L22

Comment (kehuili) [reply src/openfunction/function_context.py]: It's for runtime in FUNC_CONTEXT, so I thought it should be capitalized w/ first letter like examples here https://github.com/OpenFunction/samples/blob/bb4e664321f0a2e9d509543936abf9a9b776557c/functions-framework/golang/Knative/http/README.md#func_context

**Comment by kehuili** [comment]
> is the quick start example in the [readme](https://github.com/kehuili/functions-framework-python/tree/feature/init_binding_and_async#quickstarts) need to be updated as well?

Definitely gonna add some samples later🙆‍♀️

File: src/openfunction/function_context.py (hunk 1 above)
Comment (kehuili) [reply src/openfunction/function_context.py]: Indeed.

File: src/functions_framework/_cli.py [hunk 3]
```diff
@@ -34,10 +36,24 @@
...
 def _cli(target, source, signature_type, host, port, debug, dry_run):
-    app = create_app(target, source, signature_type)
-    if dry_run:
-        click.echo("Function: {}".format(target))
-        click.echo("URL: http://{}:{}/".format(host, port))
-        click.echo("Dry run successful, shutting down.")
+    context = _function_registry.get_openfunction_context(None)
+    
+    # determine if async or knative
+    if context and context.runtime == ASYNC_RUNTIME_TYPE:
```
Comment (benjaminhuo) [inline src/functions_framework/_cli.py]: Let's make a case insensitive comparison for xx_RUNTIME_TYPE

Comment (kehuili) [reply src/functions_framework/_cli.py]: Done.

**Review by benjaminhuo — APPROVED** [review]

**Pull request merged**